### ISSUE TITLE
Create a Plugin System

### ISSUE THREAD

**Issue by twatzl** [photoprism/photoprism#117]
I have not found any similar idea in the issues here. I hope I have not missed one, if so I am sorry.

So basically this is an idea for implementing a plugin system for Photoprism.

**What?** Basically provide the possibility to write small components which can extend the functionality of Photoprism.
Functionality could include:
- Tagging algorithms
- Generating and handling additional metadata
- Extending the search by providing additional filters (for example a search where only similar images are shown)
- Extending the information shown to the user. e.g. make a Wikipedia search based on the tags.

These are only examples. There would be endless possibilities.

**Why?** There are basically endless possibilities in the field of computer vision and for example for face detection there are multiple solutions or algorithms available. Implementing everything into Photoprism itself would make the codebase extremly complex and hard to maintain. With a plugin system Photoprism could just provide the bare core necessary to run and everything else is provided by plugins.
Also plugin developers could concentrate on their field without having to worry about loading photos, storing information, etc.
Basically it would be a separation of concerns/single responsibility thing. It would also kind of work like a microservice architecture, where every part is only responsible for a small part of the whole system, which would make it easy to swap out functionality.

**A possible way to implement this:** One way could be that plugins are just docker containers and Photoprism communicates with them using some REST interface. Why that way? Docker containers are already widely used. I know there are people who prefer not to use Docker, but it is mostly used as a kind of package manager and for that it works fine. Additionally this would solve any dependency problems, for example if some plugin requires some obscure machine learning library or something and it would allow the plugin developers to choose a language of their liking.

I am curious what you think about this, if it could be an option in the future. Maybe you also have better ideas how this could be implemented?

**Comment by lastzero** [comment]
@twatzl I agree... once we have our first stable release, we should certainly see how to make the system easily usable by developers - or maybe even for research. Right now, given the heaps of work we still have in front of us, we can't really make any detailed plans regarding plugins though. We'll keep this issue for later.

**Comment by xplosionmind** [comment]
Hello to everybody. Any news on this front? There has been no activity on this issue in three years, but I believe implementing a Plug-in system (maybe Open APIs, too?) would be a game-changer!

**Comment by lastzero** [comment]
It would, but as you can see we did almost weekly releases this year. Physically impossible to do more with our current resources. What really helps is funding so that we can grow our team. Can't pay highly qualified developers below the legal minimum wage. Merging pull requests without having time to review them isn't a winning strategy either.

**Comment by lastzero** [comment]
https://docs.photoprism.app/funding/

**Comment by xplosionmind** [comment]
> Physically impossible to do more with our current resources.

I totally understand. I wish I had more funds to provide funding or development skills to actively help, but I have neither…

Thanks a lot for your quick reply and your work, nevertheless. I really hope this can get somewhere in the future.

Thanks a lot!

Best,
T

**Comment by rob-miller** [comment]
Just an observation, if this were prioritised it would provide a route for more developers to contribute solutions to their own most important issues.  On searching solutions in this space, the biggest draw to Piwigo for me is the available ecosystem of plugins.

**Comment by ALEEF02** [comment]
Adding a plugin interface could allow for easy addition of modern-day photo features, such as any neural network applications.

**Comment by scpedicini** [comment]
I'd like to use Photoprism but am going to hold off without an easy way to write a "post-analyzer tagger" that I can connect to CLIP, Llava, etc. to add detailed tagging/exif information.

It seems like plugin system would alleviate a lot of load from maintainers, allowing developers to build the features themselves without having to have time consuming / costly PR reviews. A thriving plugin ecosystem is what makes applications like Obsidian, VS Code, etc. really stand out.
